### PR TITLE
[ELY-1617] OCSP client into trustmanager

### PR DESCRIPTION
**Pull request by hkalina**
https://issues.jboss.org/browse/ELY-1617
https://issues.jboss.org/browse/ELY-1620
https://issues.jboss.org/browse/EAP7-715

### CODE REVIEW

File: src/main/java/org/wildfly/security/asn1/ASN1Decoder.java [hunk 1]
```diff
@@ -276,6 +276,17 @@ default String decodeBMPString() throws ASN1Exception {
...
     void decodeNull() throws ASN1Exception;
 
+    /**
+     * Decode the next ASN.1 element as an enumerated.
+     *
+     * @return an integer decoded from the next element
+     * @throws ASN1Exception if the next element is not an enumerated
+     * @throws UnsupportedOperationException if not implemented
+     */
+    default BigInteger decodeEnumerated() {
```
[the remark this answers is not in the document]
Comment (hkalina) [reply src/main/java/org/wildfly/security/asn1/ASN1Decoder.java]: fixed: https://issues.jboss.org/browse/ELY-1620

**Comment by hkalina** [comment]
Should be ok now ;)

File: src/main/java/org/wildfly/security/x500/cert/ocsp/OcspChainVerifier.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    /**
+     * Check the certificate chain.
+     *
+     * @param chain the certificate chain
+     * @throws CertificateException if the validation fails (some certificate in the chain is revoked for example)
+     */
+    public void checkChain(X509Certificate[] chain) throws CertificateException {
+        for (int i = 0; i < chain.length - 1; i++) {
+            X509Certificate certificate = chain[i];
+            X509Certificate issuer = chain[i + 1];
```
Comment (jstourac) [inline src/main/java/org/wildfly/security/x500/cert/ocsp/OcspChainVerifier.java]: Possible ArrayIndexOutOfBoundsException?

Comment (hkalina) [reply src/main/java/org/wildfly/security/x500/cert/ocsp/OcspChainVerifier.java]: There is `i < chain.length - 1`

File: src/main/java/org/wildfly/security/ssl/OcspExtendedTrustManager.java [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+        } else {
+            throw new UnsupportedOperationException();
+        }
+        verifier.checkChain(chain);
+    }
+
+    @Override
+    public void checkServerTrusted(X509Certificate[] chain, String authType, Socket socket) throws CertificateException {
+        if (delegate instanceof X509ExtendedTrustManager) {
+            ((X509ExtendedTrustManager)delegate).checkClientTrusted(chain, authType, socket);
```
[the remark this answers is not in the document]
Comment (hkalina) [reply src/main/java/org/wildfly/security/ssl/OcspExtendedTrustManager.java]: +1 fixed

File: src/main/java/org/wildfly/security/ssl/OcspExtendedTrustManager.java [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+        } else {
+            throw new UnsupportedOperationException();
+        }
+        verifier.checkChain(chain);
+    }
+
+    @Override
+    public void checkServerTrusted(X509Certificate[] chain, String authType, SSLEngine sslEngine) throws CertificateException {
+        if (delegate instanceof X509ExtendedTrustManager) {
+            ((X509ExtendedTrustManager)delegate).checkClientTrusted(chain, authType, sslEngine);
```
[the remark this answers is not in the document]
Comment (hkalina) [reply src/main/java/org/wildfly/security/ssl/OcspExtendedTrustManager.java]: +1 fixed

**Comment by darranl** [comment]
Are any of the new classes in this PR intended to become public API?

**Comment by hkalina** [comment]
Only the OcspExtendedTrustManager.

**Comment by fjuma** [comment]
I've added a +1. Just a minor thing though, the last commit message should reference ELY-1617.

**Comment by jstourac** [comment]
What is status of this one? It references two ELY jiras. At least one of them is probably already superseded by https://github.com/wildfly-security/wildfly-elytron/pull/1269, correct? Not sure about the other one - https://issues.jboss.org/browse/ELY-1620. Shall we close this one? @nekdozjam

**Comment by darranl** [comment]
Closing as superseded by https://github.com/wildfly-security/wildfly-elytron/pull/1269

**Pull request closed without merging**